src/workerd/api/sockets.c++

```diff
@@ -47,52 +47,67 @@ jsg::Ref<Socket> connectImpl(
   return connectImplNoOutputLock(js, kj::mv(actualFetcher), kj::mv(address));
 }
 
-InitData initialiseSocket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) {
+InitData initialiseSocket(
+    jsg::Lock& js, kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise,
+    std::function<void()> onClose) {
   auto& context = IoContext::current();
 
   // Initialise the readable/writable streams with a custom AsyncIoStream that waits for the
   // completion of `connectionPromise` before performing reads/writes.
-  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise));
+  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise), onClose);
   auto sysStreams = newSystemMultiStream(kj::addRef(*stream), StreamEncoding::IDENTITY, context);
+  auto readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable));
+  auto writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable));
+
+  auto closeFulfiller = IoContext::current().addObject(
+      kj::heap<kj::PromiseFulfillerPair<bool>>(kj::newPromiseAndFulfiller<bool>()));
+  auto jsCloseFulfiller = IoContext::current().addObject(
+      kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()));
+  auto jsCloseFulfillerFork = jsCloseFulfiller->promise.fork();
 
   return {
-    .readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable)),
-    .writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable)),
-    .closeFulfiller = IoContext::current().addObject(
-        kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()))
+    .readable = kj::mv(readable),
+    .writable = kj::mv(writable),
+    .closeFulfiller = kj::mv(closeFulfiller),
+    .jsCloseFulfiller = kj::mv(jsCloseFulfiller),
+    .jsCloseFulfillerFork = kj::mv(jsCloseFulfillerFork)
   };
 }
 
-Socket::Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
-    Socket(initialiseSocket(kj::mv(connectionPromise))) {};
+Socket::Socket(jsg::Lock& js, kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) :
+    Socket(js, initialiseSocket(js, kj::mv(connectionPromise), [this]() { performClose(true); })) {};
+
+void Socket::close() {
+  performClose(false);
+}
 
-jsg::Promise<void> Socket::close(jsg::Lock& js) {
+void Socket::performClose(bool isImplicit) {
   if (!closeFulfiller->fulfiller->isWaiting()) {
-    return js.resolvedPromise();
+    return;
   }
 
-  auto result = js.resolvedPromise();
-  result = readable->cancel(js, nullptr);
-  result = writable->abort(js, nullptr);
-  closeFulfiller->fulfiller->fulfill();
-  return result;
+  // Readable and writable streams are closed inside callback set up in `initialiseSocket`.
+  closeFulfiller->fulfiller->fulfill(kj::mv(isImplicit));
```

Comment:
Nit: odd to `kj::mv()` a bool

Reply:
agreed and yet not `kj::mv()`'ing resulted in:

```
external/workerd/src/workerd/api/sockets.c++:90:38: error: rvalue reference to type 'bool' cannot bind to lvalue of type 'bool'
  closeFulfiller->fulfiller->fulfill(isImplicit);
                                     ^~~~~~~~~~
bazel-out/k8-dbg/bin/external/capnp-cpp/src/kj/_virtual_includes/kj-async/kj/async.h:728:28: note: passing argument to parameter 'value' here
  virtual void fulfill(T&& value) = 0;
                           ^
1 warning and 1 error generated.
```

Reply:
Yep, either `kj::mv` or `kj::cp` is in fact required here. (For a trivially-copyable type like this, either one will do exactly the same thing.)